torch/distributed/device_mesh.py

```diff
@@ -392,81 +95,123 @@ def _get_device_handle(device_type: str = "cuda"):
+                sub_stride = ceil_div(sub_stride, curr_size)
+
+            if sub_size != 1 or len(sub_res.sizes_and_strides) == 0:
+                sub_res.sizes_and_strides.append((sub_size, sub_stride * l.strides[-1]))
+
+            res.append(sub_res)
+
+        return res
+
+    # Borrowed from pycute.layout.complement
+    # https://github.com/NVIDIA/cutlass/blob/6dd13d42784ee5bfa232d2441e6b9a021c5c6290/python/pycute/layout.py#L232
+    def complement(l: Layout) -> Layout:
+        res = Layout([])
+        current_idx = 1
+
+        for size, stride in sorted(l.sizes_and_strides, key=lambda x: x[1]):
+            if stride == 0 or size == 1:
+                continue
+
+            assert current_idx <= size * stride
+
+            res.sizes_and_strides.append((stride // current_idx, current_idx))
+            current_idx = size * stride
+
+        res.sizes_and_strides.append(
+            (ceil_div(get_world_size(), current_idx), current_idx)
+        )
+
+        return coalesce(res)
+
+    def rank_tensor_to_layouts(mesh: torch.Tensor) -> list[Layout]:
+        return [Layout([(mesh.size(i), mesh.stride(i))]) for i in range(mesh.ndim)]
+
+    def layout_to_rank_group_0(l: Layout) -> list[int]:
+        return [
+            sum(c * s for c, s in zip(coord, l.strides))
+            for coord in product(*(range(s) for s in l.sizes))
+        ]
+
+    def layout_to_all_rank_groups(l: Layout) -> list[list[int]]:
+        return [
+            [outer_offset + inner_offset for inner_offset in layout_to_rank_group_0(l)]
+            for outer_offset in layout_to_rank_group_0(complement(l))
+        ]
+
+    class DeviceMeshStorage:
         def __init__(
             self,
             device_type: str,
-            mesh: Union[torch.Tensor, "ArrayLike"],
-            *,
-            mesh_dim_names: Optional[tuple[str, ...]] = None,
-            backend_override: Optional[
-                tuple[tuple[Optional[str], Optional[C10dBackend.Options]], ...]
-            ] = None,
             _init_backend: bool = True,
         ) -> None:
             self.device_type = device_type
-            if isinstance(mesh, torch.Tensor) and mesh.device.type != "cpu":
-                raise ValueError(f"`mesh` must be a CPU tensor, got {mesh}")
-            self.mesh = (
-                mesh.detach().to(dtype=torch.int)
-                if isinstance(mesh, torch.Tensor)
-                else torch.tensor(mesh, device="cpu", dtype=torch.int)
-            )
-            self.mesh_dim_names = tuple(mesh_dim_names) if mesh_dim_names else None
-            if backend_override is None:
-                backend_override = ((None, None),) * self.mesh.ndim
-
-            # private field to pre-generate DeviceMesh's hash
-            self._flatten_mesh_list = tuple(self.mesh.flatten().tolist())
-            self._thread_id = None
+            self.layouts_to_groups: dict[Layout, ProcessGroup] = {}
+            self.names_to_layouts: dict[str, Layout] = {}
```

Comment:
i'm wondering about the choice of names_to_layouts and layouts_to_groups.

What is names_to_layouts doing in the storage (cache) layer? 

if a user says "i really want to create the same PG twice" (not sure if there is a good reason, but perhaps same ranks, different settings?) then they would need a way to name the PG itself and associate it with a particular layout/mesh.

Otoh, if we want to just make this a simple cache, we could probably make it only care about layout->pg and then we keep the name-layout part inside DeviceMesh itself?

Reply:
> if a user says "i really want to create the same PG twice" (not sure if there is a good reason, but perhaps same ranks, different settings?) then they would need a way to name the PG itself and associate it with a particular layout/mesh.

I asked this question in another PR and the answer there is that if users want to use a different NCCLComm or stream, user need to create a new one themselves: https://github.com/pytorch/pytorch/pull/159482#discussion_r2268908171

Reply:
My original intention here was to make sure that names were globally unambiguous, i.e., a name always refers to the same layout, and no layout has more than one name. These principles could be up for discussion.

I do feel though that the "key" we should use to deduplicate/disambiguate ProcessGroups should be the layout, hence I'm somewhat against a direct mapping from names to PGs. Names, for me, should just be a shorthand for users to refer to a layout (since it can be hard otherwise to spell it out explicitly), for readability.